axum-extra/src/routing/typed.rs

```diff
@@ -229,6 +232,88 @@ pub trait TypedPath: std::fmt::Display {
     fn to_uri(&self) -> Uri {
         self.to_string().parse().unwrap()
     }
+
+    /// Add query parameters to a path.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use axum_extra::routing::TypedPath;
+    /// use serde::Serialize;
+    ///
+    /// #[derive(TypedPath)]
+    /// #[typed_path("/users")]
+    /// struct Users;
+    ///
+    /// #[derive(Serialize)]
+    /// struct Pagination {
+    ///     page: u32,
+    ///     per_page: u32,
+    /// }
+    ///
+    /// let path = Users.with_query_params(Pagination {
+    ///     page: 1,
+    ///     per_page: 10,
+    /// });
+    ///
+    /// assert_eq!(path.to_uri(), "/users?&page=1&per_page=10");
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// If `params` doesn't support being serialized as query params [`WithQueryParams`]'s [`Display`]
+    /// implementation will panic, and thus [`WithQueryParams::to_uri`] will also panic.
+    ///
+    /// [`WithQueryParams::to_uri`]: TypedPath::to_uri
+    /// [`Display`]: std::fmt::Display
+    fn with_query_params<T>(self, params: T) -> WithQueryParams<Self, T>
+    where
+        T: Serialize,
+        Self: Sized,
+    {
+        WithQueryParams { path: self, params }
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Copy)]
+pub struct WithQueryParams<P, T> {
+    path: P,
+    params: T,
+}
+
+impl<P, T> fmt::Display for WithQueryParams<P, T>
+where
+    P: TypedPath,
+    T: Serialize,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut out = self.path.to_string();
+        if !out.contains('?') {
+            out.push('?');
+        }
+        let mut urlencoder = form_urlencoded::Serializer::new(&mut out);
+        self.params
+            .serialize(serde_urlencoded::ser::Serializer::new(&mut urlencoder))
+            .unwrap_or_else(|err| {
+                panic!(
+                    "failed to URL encode value of type `{}`: {}",
+                    type_name::<T>(),
+                    err
+                )
+            });
```

Comment:
This panic is a little unfortunate but maybe its okay 🤔

Reply:
This only happens in the context of debugging tho right? Like we wouldn't expect this to happen in a deployment at runtime (unless a dev left a debug statement in the prod release right?) and they should be using like tracing & tracing-subscribers, right?

Reply:
This only happens in the context of debugging tho right? Like we wouldn't expect this to happen in a deployment at runtime (unless a dev left a debug statement in the prod release right?) and they should be using like tracing & tracing-subscribers, right?

Reply:
The bit that's confusing here is _when_ does a `serde_urlencoded` return an `Err("how does it error")`?
Wouldn't `encode("") -> ""`? Does it bail on certain unicode characters or something?

To me a nicer failure mode might be to default to the non-fallable `path.to_string()` eg: if we hit this error maybe just return `self` unmodified? :thinking: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9457bd28f9ac8cdd5953848e36c76553
![image](https://user-images.githubusercontent.com/10421138/218270426-2a03a243-7cea-46cc-9d90-f09b4ba02a9e.png)


Okay worked some of this out for myself. Given the failure mode of this lib, I feel like maybe serde_urlencoded is maybe _not_ the best solution. 

It shouldn't take as input things that can't be turned into a url as in the above?

Can we not make a better function that only takes structures that are able to be serialized into the url spec properly so the encoder can just return `Uri` or `String` instead of the `Result<Uri, Wtf>`?

Reply:
The bit that's confusing here is _when_ does a `serde_urlencoded` return an `Err("how does it error")`?
Wouldn't `encode("") -> ""`? Does it bail on certain unicode characters or something?

To me a nicer failure mode might be to default to the non-fallable `path.to_string()` eg: if we hit this error maybe just return `self` unmodified? :thinking: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9457bd28f9ac8cdd5953848e36c76553
![image](https://user-images.githubusercontent.com/10421138/218270426-2a03a243-7cea-46cc-9d90-f09b4ba02a9e.png)


Okay worked some of this out for myself. Given the failure mode of this lib, I feel like maybe serde_urlencoded is maybe _not_ the best solution. 

It shouldn't take as input things that can't be turned into a url as in the above?

Can we not make a better function that only takes structures that are able to be serialized into the url spec properly so the encoder can just return `Uri` or `String` instead of the `Result<Uri, Wtf>`?

Reply:
I'll refer to our discussion on the issue (https://github.com/tokio-rs/axum/issues/1740). This is because of how serde works and not serde_urlencoded's fault.